library/core/src/intrinsics.rs

```diff
@@ -1223,6 +1223,35 @@ extern "rust-intrinsic" {
     /// }
     /// ```
     ///
+    /// However, since pointers (and references) have provenance and integers do not, the following
+    /// program executes UB:
```

Comment:
I feel like this needs more comments explaining what happens -- ideally referring to some module-level docs in `std::ptr` that explain the entire provenance story.

Right now this PR is the firs to introduce the word "provenance" into stable docs.

Reply:
Also, burrying this in the examples doesn't seem appropriate. There is existing discussion of ptr2int transmutes up above and down below, it's strange that this is entirely disconnected.

Reply:
Ah, I see. I will not be the one to write the documentation for what provenance is, so we can either land this without added documentation on the basis of what Scott said here: https://github.com/rust-lang/rust/pull/121282#discussion_r1519163274 or not at all.

I'm going to force-push away the documentation from this PR.

Reply:
I'll see if I can add some wording to the docs that warn against int2ptr transmutes without talking directly about promotion. That doesn't have to block this PR.